pkg/dockerregistry/server/tagservice.go

```diff
@@ -0,0 +1,206 @@
+			continue
+		}
+		if !t.repo.pullthrough {
+			managed, found := managedImages[history.Items[0].Image]
+			if !found {
+				dgst, err := digest.ParseDigest(history.Items[0].Image)
+				if err != nil {
+					context.GetLogger(ctx).Errorf("bad digest %s: %v", history.Items[0].Image, err)
+					continue
+				}
+
+				image, err := t.repo.getImage(dgst)
+				if err != nil {
+					context.GetLogger(ctx).Errorf("unable to get image %s/%s %s: %v", t.repo.namespace, t.repo.name, dgst.String(), err)
+					continue
+				}
+				managed = isImageManaged(image)
+				managedImages[history.Items[0].Image] = managed
+			}
+
+			if !managed {
+				continue
+			}
+		}
+		tags = append(tags, tag)
+	}
+	return tags, nil
+}
+
+func (t tagService) Lookup(ctx context.Context, desc distribution.Descriptor) ([]string, error) {
+	tags := []string{}
+
+	imageStream, err := t.repo.getImageStream()
+	if err != nil {
+		context.GetLogger(ctx).Errorf("error retrieving ImageStream %s/%s: %v", t.repo.namespace, t.repo.name, err)
+		return tags, distribution.ErrRepositoryUnknown{Name: t.repo.Named().Name()}
+	}
+
+	managedImages := make(map[string]bool)
+
+	for tag, history := range imageStream.Status.Tags {
+		if len(history.Items) == 0 {
+			continue
+		}
+
+		dgst, err := digest.ParseDigest(history.Items[0].Image)
+		if err != nil {
+			context.GetLogger(ctx).Errorf("bad digest %s: %v", history.Items[0].Image, err)
+			continue
+		}
+
+		if dgst != desc.Digest {
+			continue
+		}
+
+		if !t.repo.pullthrough {
+			managed, found := managedImages[history.Items[0].Image]
+			if !found {
+				image, err := t.repo.getImage(dgst)
+				if err != nil {
+					context.GetLogger(ctx).Errorf("unable to get image %s/%s %s: %v", t.repo.namespace, t.repo.name, dgst.String(), err)
+					continue
+				}
+				managed = isImageManaged(image)
+				managedImages[history.Items[0].Image] = managed
+			}
+
+			if !managed {
+				continue
+			}
+		}
+		tags = append(tags, tag)
+	}
+
+	return tags, nil
+}
+
+func (t tagService) Tag(ctx context.Context, tag string, dgst distribution.Descriptor) error {
+	imageStream, err := t.repo.getImageStream()
+	if err != nil {
+		context.GetLogger(ctx).Errorf("error retrieving ImageStream %s/%s: %v", t.repo.namespace, t.repo.name, err)
+		return distribution.ErrRepositoryUnknown{Name: t.repo.Named().Name()}
+	}
+
+	image, err := t.repo.registryOSClient.Images().Get(dgst.Digest.String())
+	if err != nil {
+		context.GetLogger(ctx).Errorf("unable to get image: %s", dgst.Digest.String())
+		return err
+	}
+	image.SetResourceVersion("")
+
+	if t.repo.pullthrough && !isImageManaged(image) {
+		return distribution.ErrRepositoryUnknown{Name: t.repo.Named().Name()}
+	}
+
+	ism := imageapi.ImageStreamMapping{
+		ObjectMeta: kapi.ObjectMeta{
+			Namespace: imageStream.Namespace,
+			Name:      imageStream.Name,
+		},
+		Tag:   tag,
+		Image: *image,
+	}
+
+	return t.repo.registryOSClient.ImageStreamMappings(imageStream.Namespace).Create(&ism)
```

Comment:
Ahh. you're right about not creating the image stream. After all, you've just verified it actually exists at the method's beginning. I don't think this will be invoked so often that we need to care about race with IS delete. You should, however, handle quota abuse.